### PR TITLE
Graduate metrics/v1alpha1 to v1beta1

### PR DESCRIPTION
**Pull request by DirectXMan12**
This introduces v1beta1 of the resource metrics API, previously in alpha.
The v1alpha1 version remains for compatibility with the Heapster legacy version
of the resource metrics API, which is compatible with the v1alpha1 version.  It also
renames the v1beta1 version to `resource-metrics.metrics.k8s.io`.

The HPA controller's REST clients (but not the legacy client) have been migrated as well.

Part of kubernetes/features#118.

```release-note
Migrate the metrics/v1alpha1 API to metrics/v1beta1.  The HorizontalPodAutoscaler
controller REST client now uses that version.  For v1beta1, the API is now known as
resource-metrics.metrics.k8s.io.
```

### CODE REVIEW

**Comment by DirectXMan12** [comment]
cc @kubernetes/sig-instrumentation-api-reviews 
@kubernetes/sig-autoscaling-api-reviews 

cc @piosz

**Comment by smarterclayton** [comment]
Is there a feature issue to attach to this?

**Comment by DirectXMan12** [comment]
Whoops, yep, will do -- it's part of kubernetes/features#118 ("the master metrics API... needs to be stabilized").

**Comment by DirectXMan12** [comment]
let's see if that actually kicks the tests off -- it's been sitting here for a day, not doing anything

**Comment by smarterclayton** [comment]
So other than the name of the group I don't see any issues here with API.  Will need sig approval and LGTM though.

File: staging/src/k8s.io/metrics/pkg/apis/metrics/register.go [hunk 1]
```diff
@@ -22,7 +22,7 @@ import (
 )
 
 // GroupName is the group name use in this package
-const GroupName = "metrics"
+const GroupName = "resource-metrics.metrics.k8s.io"
```
Comment (piosz) [inline staging/src/k8s.io/metrics/pkg/apis/metrics/register.go]: Why not having just `metrics.k8s.io`. metrics.metrics sounds weird to me.

Comment (DirectXMan12) [reply staging/src/k8s.io/metrics/pkg/apis/metrics/register.go]: We already have custom-metrics.metrics.k8s.io, so it just seemed strange to use it as both a "namespace" for groups and a group itself.  This would make it resource-metrics.metrics and custom-metrics.metrics, so they match.

Comment (piosz) [reply staging/src/k8s.io/metrics/pkg/apis/metrics/register.go]: can we have:
`metrics.k8s.io` for resource metrics
`custom.metrics.k8s.io` for custom metrics?

**Comment by smarterclayton** [comment]
Yeah those groups sound fine to me.

**Comment by smarterclayton** [comment]
We can get a release exception if that goes past code freeze since this would be new APIs, so if we generally agree on the shorter names we should do that for 1.8

File: staging/src/k8s.io/metrics/pkg/apis/metrics/doc.go [hunk 2]
```diff
@@ -14,8 +14,6 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-// TODO: this needs to be cross-registered as metrics.metrics.k8s.io
-
 // +k8s:deepcopy-gen=package,register
-
+// +groupName=resource-metrics.metrics.k8s.io
```
[the remark this answers is not in the document]
Comment (smarterclayton) [reply staging/src/k8s.io/metrics/pkg/apis/metrics/doc.go]: (this was my old comment that was not committed).

I'd prefer `metrics.k8s.io` and `custom.metrics.k8s.io`

**Comment by piosz** [comment]
@smarterclayton which group names sound fine for you?

**Comment by smarterclayton** [comment]
> metrics.k8s.io for resource metrics
> custom.metrics.k8s.io for custom metrics?

SGTM

**Comment by piosz** [comment]
@smarterclayton cool
@DirectXMan12 WDYT?

File: pkg/apis/batch/v1beta1/zz_generated.conversion.go [hunk 3]
```diff
@@ -215,7 +214,8 @@ func Convert_batch_JobTemplate_To_v1beta1_JobTemplate(in *batch.JobTemplate, out
 
 func autoConvert_v1beta1_JobTemplateSpec_To_batch_JobTemplateSpec(in *v1beta1.JobTemplateSpec, out *batch.JobTemplateSpec, s conversion.Scope) error {
 	out.ObjectMeta = in.ObjectMeta
-	if err := batch_v1.Convert_v1_JobSpec_To_batch_JobSpec(&in.Spec, &out.Spec, s); err != nil {
+	// TODO: Inefficient conversion - can we improve it?
```
Comment (piosz) [inline pkg/apis/batch/v1beta1/zz_generated.conversion.go]: Is this change intended? Looks weird.

File: pkg/generated/openapi/BUILD [hunk 4]
```diff
@@ -59,6 +59,7 @@ openapi_library(
         "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1beta1",
         "k8s.io/metrics/pkg/apis/custom_metrics/v1alpha1",
         "k8s.io/metrics/pkg/apis/metrics/v1alpha1",
```
Comment (piosz) [inline pkg/generated/openapi/BUILD]: shouldn't this be removed?

File: staging/src/k8s.io/code-generator/cmd/go-to-protobuf/protobuf/cmd.go [hunk 5]
```diff
@@ -98,6 +98,7 @@ func New() *Generator {
 			`k8s.io/api/networking/v1`,
 			`k8s.io/kubernetes/federation/apis/federation/v1beta1`,
 			`k8s.io/metrics/pkg/apis/metrics/v1alpha1`,
```
Comment (piosz) [inline staging/src/k8s.io/code-generator/cmd/go-to-protobuf/protobuf/cmd.go]: same here

File: staging/src/k8s.io/metrics/hack/update-codegen.sh [hunk 6]
```diff
@@ -36,7 +36,7 @@ PREFIX=k8s.io/metrics/pkg/apis
 INPUT_BASE="--input-base ${PREFIX}"
 CLIENTSET_PATH="--clientset-path k8s.io/metrics/pkg/client/clientset_generated"
 
-${CLIENTGEN} --clientset-name="clientset" ${INPUT_BASE} --input metrics/v1alpha1 ${CLIENTSET_PATH} --output-base ${SCRIPT_BASE}
+${CLIENTGEN} --clientset-name="clientset" ${INPUT_BASE} --input metrics/v1alpha1 --input metrics/v1beta1 ${CLIENTSET_PATH} --output-base ${SCRIPT_BASE}
```
Comment (piosz) [inline staging/src/k8s.io/metrics/hack/update-codegen.sh]: remove `metrics/v1alpha1`?

File: staging/src/k8s.io/metrics/pkg/apis/metrics/doc.go (hunk 2 above)
Comment (piosz) [inline staging/src/k8s.io/metrics/pkg/apis/metrics/doc.go]: just `metrics.k8s.io`

**Comment by DirectXMan12** [comment]
Names sound fine to me.  Will fix when I get home from Flock later today.

**Comment by DirectXMan12** [comment]
(although in the true spirit of bikeshedding, I'd prefer resource.metrics.k8s.io, for consistency)

**Comment by DirectXMan12** [comment]
FYI, we have to preserve v1alpha1 until we remove the legacy metrics client from the HPA

File: pkg/apis/batch/v1beta1/zz_generated.conversion.go (hunk 3 above)
Comment (DirectXMan12) [reply pkg/apis/batch/v1beta1/zz_generated.conversion.go]: no, it's not.  It's weird that the conversion-gen changed it...

**Comment by piosz** [comment]
Let's go with `metrics.k8s.io` and `custom.metrics.k8s.io` then. Otherwise LGTM. Ping me once you update the code, so I'll apply the label.

> FYI, we have to preserve v1alpha1 until we remove the legacy metrics client from the HPA

Oh I see. Please disregard my comments about v1alpha1.

**Comment by piosz** [comment]
I still can see modified `pkg/apis/batch/v1beta1/zz_generated.conversion.go`

File: staging/src/k8s.io/metrics/pkg/apis/metrics/v1beta1/register.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+package v1beta1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// GroupName is the group name use in this package
+const GroupName = "metrics"
```
Comment (smarterclayton) [inline staging/src/k8s.io/metrics/pkg/apis/metrics/v1beta1/register.go]: metrics.k8s.io?

**Comment by smarterclayton** [comment]
/approve

Names `custom.metrics.k8s.io` and `metrics.k8s.io`.  The other APIs you are promoting look fine.

**Comment by sjenning** [comment]
Wait... when did kops drop off the required list?!  I'm not complaining; it was blocking the world.  Just didn't notice before I started the retest.

**Comment by liggitt** [comment]
It is non-blocking while they work out why it wasn't running tests using the test code from the PR being tested

**Comment by smarterclayton** [comment]
@piosz is there an active release exception covering this yet?

This is API approved

**Comment by piosz** [comment]
@smarterclayton I don't know how the exception process works but my feeling is that it was reviewed and approved before the code freeze.

**Comment by piosz** [comment]
Ups. I unintentionally triggered merge. We can consider reverting this if people disagree with my https://github.com/kubernetes/kubernetes/pull/51653#issuecomment-327420664.
